event-svc/src/event/validator/event.rs

```diff
@@ -0,0 +1,187 @@
+                key_cid,
+                cid
+            )));
+        }
+
+        Ok(Self {
+            key: item.key.to_owned(),
+            cid: key_cid,
+            event: Arc::new(event),
+        })
+    }
+}
+
+impl ValidatedEvents {
+    pub fn new_with_expected_valid(valid: usize) -> Self {
+        // sort of arbitrary sizes, not betting on invalid events
+        Self {
+            valid: Vec::with_capacity(valid),
+            pending: Vec::with_capacity(valid / 4),
+            invalid: Vec::new(),
+        }
+    }
+
+    pub fn extend_with(&mut self, other: Self) {
+        self.valid.extend(other.valid);
+        self.invalid.extend(other.invalid);
+        self.pending.extend(other.pending);
+    }
+}
+
+#[derive(Debug)]
+pub struct EventValidator<'a> {
+    pool: &'a SqlitePool,
+    /// Whether to require the init event exists, and if we should check the the signature against
+    /// the init controller. True in all cases except a data migration.
+    require_init_event: bool,
+    /// Whether we should check the signature is currently valid or simply whether it was once valid
+    check_exp: bool,
+}
+
+impl<'a> EventValidator<'a> {
+    fn new(pool: &'a SqlitePool, deliverable_req: DeliverableRequirement) -> Self {
+        let (require_init_event, check_exp) = match deliverable_req {
+            DeliverableRequirement::Immediate => (true, true),
+            DeliverableRequirement::Asap => (true, false),
+            DeliverableRequirement::Lazy => (false, false),
+        };
+        Self {
+            pool,
+            require_init_event,
+            check_exp,
+        }
+    }
+
+    pub(crate) async fn validate_events(
+        pool: &'a SqlitePool,
+        deliverable_req: DeliverableRequirement,
+        parsed_events: Vec<UnvalidatedEvent>,
+    ) -> Result<ValidatedEvents> {
+        let validator = Self::new(pool, deliverable_req);
+
+        let mut validated = ValidatedEvents::new_with_expected_valid(parsed_events.len());
+        // partition the events by type of validation needed and delegate to validators
+        let grouped = GroupedEvents::from(parsed_events);
+
+        let (validated_signed, validated_time) = try_join!(
+            validator.validate_signed_events(grouped.signed_events),
+            validator.validate_time_events(grouped.time_events)
+        )?;
+
+        validated.extend_with(validated_signed);
+        validated.extend_with(validated_time);
+
+        if !validated.invalid.is_empty() {
+            tracing::warn!(count=%validated.invalid.len(), "invalid events discovered");
+        }
+        Ok(validated)
+    }
+
+    async fn validate_signed_events(&self, events: SignedEvents) -> Result<ValidatedEvents> {
+        let opts = if self.check_exp {
+            ceramic_validation::VerifyJwsOpts::default()
+        } else {
+            ceramic_validation::VerifyJwsOpts {
+                at_time: ceramic_validation::AtTime::SkipTimeChecks,
```

Comment:
I'm not sure.. do we want to throw out events discovered on the network that have expired cacaos or allow them and then let the conclusion be that they're expired? I think I'd prefer to ignore time information from recon events because it will add an entire new level of ordering to validation that might be a more natural fit for aggregation/conclusions.

Also for did:key, we don't want to check if the DID is expired so I think skip (or some variation on it) will probably make sense for recon discovered data.

Reply:
I still don't see why recon is any different from the API with regard to CACAO expiration.  Either the event is valid or it isn't, and the logic for determining if it's valid depends on timestamps from anchors.

Also I don't see how the existence of did:key signed events factors into this?

Reply:
did:keys can also expire, be rotated or revoked, though the exact fields to check seem a bit dependent on your use case. We check `nextUpdate` in [js-did](https://github.com/ceramicnetwork/js-did/blob/3a8ad6d2c743b9b22c067961ae0e71818ff76443/packages/dids/src/did.ts#L353).

the rules should be the same, but the current flow is the problem.. one is happening in parallel (or in the future) and the other is being applied to the stream tip. also, if an event is invalid because a cacao is expired when we find it on the network, should we drop it or store it? in the future we *may* discover the time event that makes it valid, so drawing a conclusion that it's not valid as part of the stream makes sense, but rejecting it completely seems a bit excessive.